### ISSUE TITLE
Print race condition

### ISSUE THREAD

**Issue by rwols** [tomv564/LSP#557]
The server prints and LSP prints may intertwine. Here is a sample clangd session from the console:

```
LSP: <--  textDocument/publishDiagnostics
LSP:      {'diagnostics': [{'message': "Unexpected namespace name 'std': expected expression", 'range': {'end': {'line': 32, 'character': 11}, 'start': {'line': 32, 'character': 8}}, 'severity': 1}], 'uri': 'file:///Users/raoulwols/dev/qi.system/libs/common/test/test_hillclimb.cpp'}
server: I[23:01:28.472] --> reply:textDocument/documentHighlight(19) 33 ms
LSP: severity check 1 <= 3
LSP:      []
LSP:  --> textDocument/codeAction
server: I[23:01:28.741] <-- textDocument/codeAction(20)
LSP:server:      []
 I[23:01:28.741] --> reply:textDocument/codeAction(20) 0 ms
LSP:  --> textDocument/didChange
server: I[23:01:29.739] <-- textDocument/didChange
LSP:  --> textDocument/codeAction
server: I[23:01:30.049] <-- textDocument/codeAction(21)
server: I[23:01:30.049] --> reply:textDocument/codeAction(21) 0 ms
LSP:      []
server: I[23:01:30.240] Updating file /Users/raoulwols/dev/qi.system/libs/common/test/test_hillclimb.cpp with command [/Users/raoulwols/dev/qi.system/build/system-debug-shared] /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -DSTLAB_FORCE_BOOST_OPTIONAL -Iinclude -I../../libs/common/include -I../../libs/testing/include -isystem /usr/local/include -isystem /usr/local/include/opencv -g -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk -Werror=implicit-function-declaration -Werror=return-type -std=gnu++14 -o libs/common/test/CMakeFiles/test_common.dir/test_hillclimb.cpp.o -c /Users/raoulwols/dev/qi.system/libs/common/test/test_hillclimb.cpp -resource-dir=/usr/local/bin/../lib/clang/8.0.0
server: I[23:01:30.273] --> textDocument/publishDiagnostics
LSP: <--  textDocument/publishDiagnostics
LSP:      {'diagnostics': [{'message': 'Expected unqualified-id', 'range': {'end': {'line': 34, 'character': 14}, 'start': {'line': 34, 'character': 8}}, 'severity': 1}], 'uri': 'file:///Users/raoulwols/dev/qi.system/libs/common/test/test_hillclimb.cpp'}
LSP: severity check 1 <= 3
```

notice in particular the line `LSP:server:      []`.

**Comment by tomv564** [comment]
Is this a common end-user issue? 
This feels like a complication for an inconvenience to developers. Without testing impact, I feel like logging is the last place I'd want performance issues to come from.

**Comment by haferburg** [comment]
The change is to lock a mutex before calling `print`. There is no risk of contention or deadlocks, `print` is not going to block for a long time. End users have `log_debug` and `log_exceptions` disabled, so only `server_log` is affected. `server_log` is only called for `"window/logMessage"`, and used via `attach_logger` if `log_stderr` is set. I think this is just the right thing to do if you use multithreading.

**Comment by rwols** [comment]
will by fixed by #811
